src/libdredd/src/mutation_replace_expr.cc

```diff
@@ -62,23 +62,39 @@ std::string MutationReplaceExpr::GetFunctionName(
     result += "_lvalue";
   }
 
+  if (IsRedundantOperatorInsertion(expr_, ast_context)) {
+    result += "_optimised";
+  }
+
   return result;
 }
 
 bool MutationReplaceExpr::ExprIsEquivalentTo(const clang::Expr& expr,
                                              int constant,
                                              clang::ASTContext& ast_context) {
-  clang::Expr::EvalResult eval_result;
-  if (expr.EvaluateAsInt(eval_result, ast_context)) {
-    return llvm::APSInt::isSameValue(eval_result.Val.getInt(),
+  clang::Expr::EvalResult int_result;
+  if (expr.getType()->isIntegerType() &&
+      expr.EvaluateAsInt(int_result, ast_context)) {
+    return llvm::APSInt::isSameValue(int_result.Val.getInt(),
                                      llvm::APSInt::get(constant));
   }
   return false;
 }
 
+bool MutationReplaceExpr::IsRedundantOperatorInsertion(
+    const clang::Expr& expr, clang::ASTContext& ast_context) {
+  bool result;
+  return (expr.getType()->isBooleanType() &&
+          expr.EvaluateAsBooleanCondition(result, ast_context)) ||
```

Comment:
Add a comment that we use `result` to capture that the expression evaluates to if it indeed evaluates to a constant, but that we don't actually use this value because either way it means the operator is redundant.

Separate comment: I would change it from `result` to something else, because `result` typically refers to the value that the current method will eventually return. I'd call it `unused_bool_value` or something.

Reply:
Done.